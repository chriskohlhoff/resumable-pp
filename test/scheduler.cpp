#include <stdio.h>

class function_base
{
public:
  virtual ~function_base() {}
  virtual void invoke() = 0;

private:
  friend class scheduler;
  function_base* next_ = nullptr;
};

template <class F>
class function : public function_base
{
public:
  explicit function(F f)
    : f_(f)
  {
  }

  virtual ~function()
  {
  }

  virtual void invoke()
  {
    f_();
  }

private:
  F f_;
};

class scheduler
{
public:
  template <class F>
  int post(F f)
  {
    if (last_)
    {
      last_->next_ = new function<F>(f);
      last_ = last_->next_;
    }
    else
    {
      first_ = last_ = new function<F>(f);
    }

    return 0;
  }

  void run()
  {
    while (first_)
    {
      function_base* f = first_;
      first_ = first_->next_;
      if (!first_)
        last_ = nullptr;
      f->invoke();
      delete f;
    }
  }

private:
  function_base* first_ = nullptr;
  function_base* last_ = nullptr;
};

int main()
{
  scheduler sched;

  auto f = [&sched, i = int(0)]() resumable
  {
    for (i = 0; i < 10; ++i)
    {
      printf("f: %d\n", i);
      yield sched.post(*lambda_this);
    }
  };

  auto g = [&sched, i = int(0)]() resumable
  {
    for (i = 0; i < 5; ++i)
    {
      printf("g: %d\n", i);
      yield sched.post(*lambda_this);
    }
  };

  sched.post(f);
  sched.post(g);
  sched.run();
}
