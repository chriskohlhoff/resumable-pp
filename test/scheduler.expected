f: 0
g: 0
f: 1
g: 1
f: 2
g: 2
f: 3
g: 3
f: 4
g: 4
f: 5
f: 6
f: 7
f: 8
f: 9
